pycfmodel/model/resources/properties/statement.py

```diff
@@ -12,61 +12,85 @@
+
+logger = logging.getLogger(__file__)
+
+PrincipalTypes = Union[ResolvableStr, List[ResolvableStr], Dict[str, Union[ResolvableStr, List[ResolvableStr]]]]
+
+
+class Statement(Property):
+    Sid: Optional[ResolvableStr] = None
+    Effect: Optional[ResolvableStr] = None
+    Condition: Optional[ResolvableStr] = None
+    Principal: Optional[PrincipalTypes] = None
+    NotPrincipal: Optional[PrincipalTypes] = None
+    Action: Optional[ResolvableStrOrList] = None
+    NotAction: Optional[ResolvableStrOrList] = None
+    Resource: Optional[ResolvableStrOrList] = None
+    NotResource: Optional[ResolvableStrOrList] = None
+
+    def get_action_list(self) -> List[ResolvableStr]:
+        action_list = []
+        for actions in [self.Action, self.NotAction]:
+            if isinstance(actions, List):
+                action_list.extend(actions)
+            elif isinstance(actions, (str, dict)):
+                action_list.append(actions)
+        return action_list
+
+    def get_resource_list(self) -> List[ResolvableStr]:
+        resource_list = []
+        for resources in [self.Resource, self.NotResource]:
+            if isinstance(resources, List):
+                resource_list.extend(resources)
+            elif isinstance(resources, (str, dict)):
+                resource_list.append(resources)
+        return resource_list
+
+    def get_principal_list(self) -> List[ResolvableStr]:
+        principal_list = []
+        for principals in [self.Principal, self.NotPrincipal]:
+            if isinstance(principals, list):
+                principal_list.extend(principals)
+            elif isinstance(principals, str):
+                principal_list.append(principals)
+            elif is_resolvable_dict(principals):
+                principal_list.append(principals)
+            elif isinstance(principals, dict):
+                for value in principals.values():
+                    if isinstance(value, (str, Dict)):
+                        principal_list.append(value)
+                    elif isinstance(value, List):
+                        principal_list.extend(value)
+            elif principals is not None:
+                raise ValueError(f"Not supported type: {type(principals)}")
+        return principal_list
+
+    def actions_with(self, pattern: Pattern) -> List[str]:
+        return [action for action in self.get_action_list() if isinstance(action, str) and pattern.match(action)]
+
+    def principals_with(self, pattern: Pattern) -> List[str]:
+        return [
+            principal
+            for principal in self.get_principal_list()
+            if isinstance(principal, str) and pattern.match(principal)
+        ]
+
+    def resources_with(self, pattern: Pattern) -> List[str]:
+        return [
+            resource for resource in self.get_resource_list() if isinstance(resource, str) and pattern.match(resource)
+        ]
+
+    def non_whitelisted_principals(self, whitelist: List[str]) -> List[str]:
```

Comment:
This method doesn't seem like it should be in pycfmodel, more of a cfripper thing

Reply:
Also the previous method had         
```
if not self.principal or self.condition:
            return []	
```
This one should take condition into account too if we want to keep it

Reply:
Because we are ignoring a part of the model. This method is to return all non whitelisted principals for a statement. Not just the ones that are in `Principal`